src/Raven.Server/Documents/Indexes/VectorSearch/GenerateEmbeddings.cs

```diff
@@ -1,28 +1,71 @@
+#pragma warning disable SKEXP0070
 using System;
+using System.Collections.Generic;
 using System.IO;
+using System.Linq;
+using System.Reflection;
 using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
+using System.Text;
+using FastBertTokenizer;
+using Microsoft.ML.OnnxRuntime;
+using Microsoft.SemanticKernel.Connectors.Onnx;
 using Raven.Client.Documents.Indexes.Vector;
 using Raven.Client.Documents.Queries.Vector;
-using SmartComponents.LocalEmbeddings;
+using Raven.Server.Config;
 using Sparrow;
 using Sparrow.Server;
+using InvalidOperationException = System.InvalidOperationException;
 using VectorValue = Corax.Utils.VectorValue;
 
 namespace Raven.Server.Documents.Indexes.VectorSearch;
 
 public static class GenerateEmbeddings
 {
+    private static readonly ConstructorInfo BertOnnxTextEmbeddingGenerationServiceCtor;
+
     // Dimensions (buffer size) from internals of SmartComponents.
     private const int F32Size = 1536;
-    private static readonly LocalEmbedder Embedder = new();
+
+    private static SessionOptions OnnxSessionOptions;
+
+    internal static readonly Lazy<BertOnnxTextEmbeddingGenerationService> Embedder = new(CreateTextEmbeddingGenerationService);
+
+    static GenerateEmbeddings()
+    {
+        BertOnnxTextEmbeddingGenerationServiceCtor = typeof(BertOnnxTextEmbeddingGenerationService).GetConstructor(BindingFlags.NonPublic | BindingFlags.Instance, [typeof(InferenceSession), typeof(BertTokenizer), typeof(int), typeof(BertOnnxOptions)]);
+        if (BertOnnxTextEmbeddingGenerationServiceCtor == null)
+            throw new InvalidOperationException($"Could not find constructor for {typeof(BertOnnxTextEmbeddingGenerationService)}.");
+    }
+
+    public static void Configure(RavenConfiguration configuration)
+    {
+        if (Embedder.IsValueCreated)
+            throw new InvalidOperationException("Embedder has already been initialized.");
+        
+        var numberOfCoresToUse = CalculateNumberOfCoresForOnnx(configuration);
+
+        OnnxSessionOptions = new SessionOptions() { IntraOpNumThreads = numberOfCoresToUse };
+    }
+    
+    private static int CalculateNumberOfCoresForOnnx(RavenConfiguration configuration)
+    {
+        var cores = (int)MathF.Floor(Environment.ProcessorCount * ((float)configuration.Indexing.MaxPercentageOfThreadsForEmbeddings / 100));
+            
+        var numberOfCoresToUse = int.Max(1, cores);
+        
+        return numberOfCoresToUse;
+    }
```

Comment:
I think it would be _much_ safer_ to be able to dedicate an exact number of cores, rather than going with %

Reply:
Default can be

```
var cores = Environment.ProcessorCount switch
  <= 2 => 1,
  <= 8 => 2,
  <= 16 => 4,
  _ => 6
```

Reply:
We can do that, but e.g. the values you have will hit the perf significantly. On my laptop switching from all cores to 25% hits the perf by 60%, thought that percentage based could be better for high core machines, your call.

Reply:
spoke with @ayende , lets go with CPU based one, not % based